### DISCUSSION TITLE
`IllegalStateException`

### DISCUSSION THREAD

**Discussion by ShahOdin** [http4s/http4s#7502] [Q&A] [answered]
The following code: 

```scala
import cats.effect.{IO, IOApp}
import cats.effect.IO.asyncForIO
import cats.effect.kernel.Resource
import org.http4s.blaze.client.BlazeClientBuilder
import org.http4s.client.Client

import scala.concurrent.duration.*

object Demo extends IOApp.Simple:

  val clientResource: Resource[IO, Client[IO]] = BlazeClientBuilder[IO].resource

  def query: IO[Int] = clientResource.use(
    _
      .statusFromString(s"https://postman-echo.com/get")
      .map(_.code)
      .flatTap(s => IO.println(s"request: returned with status: $s"))
  )

  override def run: IO[Unit] =
    fs2.Stream
      .range(1, 100)
      .metered(1.seconds)
      .evalMap(_ => query)
      .compile
      .drain
```

throws the following logs: 

> java.lang.IllegalStateException: supervisor already shutdown
	at get @ fs2.internal.Scope.openScope(Scope.scala:275)
	at get @ fs2.internal.Scope.openScope(Scope.scala:275)
	at unique @ fs2.Compiler$Target$ConcurrentTarget.unique(Compiler.scala:194)

Note that you also see after each request:

> INFO  org.http4s.blaze.client.PoolManager - Shutting down connection pool: curAllocated=1 idleQueues.size=1 waitQueue.size=0 maxWaitQueueLimit=256 closed=false

if I change the code to: 

```scala
import cats.effect.{IO, IOApp, Resource}
import cats.effect.IO.asyncForIO
import org.http4s.blaze.client.BlazeClientBuilder
import org.http4s.client.Client

import scala.concurrent.duration.*

object Demo extends IOApp.Simple:

  val clientResource: Resource[IO, Client[IO]] = BlazeClientBuilder[IO].resource

  def query(client: Client[IO]): IO[Int] = client
    .statusFromString(s"https://postman-echo.com/get")
    .map(_.code)
    .flatTap(s => IO.println(s"request: returned with status: $s"))

  override def run: IO[Unit] =
    clientResource.use(c =>
      fs2.Stream
        .range(1, 100)
        .metered(1.seconds)
        .evalMap(_ => query(c))
        .compile
        .drain
    )
```

the error logs go away. and the connection pool isn't closed after every request either. 

Is this a bug or a feature? Why is it happening?

**Comment by armanbilge** [comment]
In the first example, `query` is creating a new HTTP client every time it runs.

In the second example, you create the client once at the beginning of the program and that single client is shared for every `query`. This is the preferred way, since as you note the connection pool can be shared.

**Comment by ShahOdin** [comment]
Thanks Arman. 
Sure, but best practices aside, how do we explain the supervisor shutdown etc.

**Reply by armanbilge** [reply, marked as answer]
> how do we explain the supervisor shutdown etc.

My guess would be that it is a bug in [http4s/blaze](https://github.com/http4s/blaze). Personally I recommend using Ember or another backend such as the JDK client or Netty.

https://http4s.org/v0.23/docs/integrations.html

**Comment by ShahOdin** [comment]
Can confirm using the ember client resolved the issue. 

<details>
  <summary>Click to see code</summary>

  ```scala
import cats.effect.{IO, IOApp, Resource}
import cats.effect.IO.asyncForIO
import org.http4s.ember.client.EmberClientBuilder
import org.http4s.client.Client

import scala.concurrent.duration.*

  object Demo extends IOApp.Simple:

  val clientResource: Resource[IO, Client[IO]] = EmberClientBuilder.default[IO].build

  def query: IO[Int] = clientResource.use(
    _
      .statusFromString(s"https://postman-echo.com/get")
      .map(_.code)
      .flatTap(s => IO.println(s"request: returned with status: $s"))
  )

  override def run: IO[Unit] =
    fs2.Stream
      .range(1, 100)
      .metered(1.seconds)
      .evalMap(_ => query)
      .compile
      .drain
```
</details>

Would you suggest I raise it as a bug for https://github.com/http4s/blaze

**Reply by armanbilge** [reply]
> Would you suggest I raise it as a bug for

Sure, that would be great, thanks!

**Reply by ShahOdin** [reply]
done: https://github.com/http4s/blaze/issues/894
